File: sudokusolver.py
import sys
import collections
import csv

def readgridfile(fileName):
    '''Read grid from grid.txt and output as list[list[int]]
    rtype: list[list[int]]
    '''
    with open(fileName) as file:
        board = [list(map(int, num)) for num in csv.reader(file)]
    return board

def isValid(board):
    '''Checks if current board is a valid board before attempting to solve.
    rtype: bool
    '''
    if len(board) != 9 and len(board[0]) != 9:
        return False
    
    cols = collections.defaultdict(set)
    rows = collections.defaultdict(set)
    squares = collections.defaultdict(set)

    for r in range(9):
        for c in range(9):
            if board[r][c] == 0:
                continue
            
            if (board[r][c] in rows[r] or
                board[r][c] in cols[c] or
                board[r][c] in squares[r // 3, c // 3]):
                return False
            
            cols[c].add(board[r][c])
            rows[r].add(board[r][c])
            squares[(r // 3, c // 3)].add(board[r][c])
    
    return True

def solveSudoku(board):
    if isComplete(board):
        return board  # Puzzle is solved
    
    row, col = findNextEmptyCell(board)
    
    for num in range(1,10):
        if isValidMove(board, row, col, num):
            board[row][col] = num
            
            if solveSudoku(board) is not None:
                return board  # Successfully solved
            
            board[row][col] = 0  # Backtrack
        
    return None  # No solution found

def isComplete(board):
    for row in board:
        for cell in row:
            if cell == 0:
                return False
    return True

def findNextEmptyCell(board):
    for row in range(9):
        for col in range(9):
            if board[row][col] == 0:
                return row, col
    return -1, -1  # No empty cell found

def isValidMove(board, row, col, num):
    # Check row and column
    for i in range(9):
        if board[row][i] == num or board[i][col] == num:
            return False
    
    # Check 3x3 subgrid
    startRow, startCol = row - row % 3, col - col % 3
    for i in range(3):
        for j in range(3):
            if board[startRow + i][startCol + j] == num:
                return False
    
    return True
    

def main():
    if len(sys.argv) >= 3 or len(sys.argv) <= 1:
        print("Usage: sudokusolver.py filepath")
        return -1
    
    board = readgridfile(sys.argv[1])
    # print(board)

    if not isValid(board):
        print("Board is not valid")
        return -1
    
    solvedBoard = solveSudoku(board)
    for row in solvedBoard:
        print(row)
        
if __name__ == "__main__":
    main()

File: sudoku-solver.js
let cols = Array.from({ length: 9 }, () => new Set());
let rows = Array.from({ length: 9 }, () => new Set());
let squares = Array.from({ length: 9 }, () => new Set());

const sudoku_board = document.querySelector('.sudoku-board');

for (let i = 0; i < 81; i++) {
    const input = document.createElement('input');
    input.type = 'text';
    input.maxLength = 1;
    sudoku_board.appendChild(input);
}

const cells = document.querySelectorAll('.sudoku-board input');
cells.forEach(cell => {
    cell.addEventListener('input', (e) => {
        const value = e.target.value;
        if (!/^[1-9]$/.test(value)) {
            e.target.value = '';
        }
    });
});

function resetSets() {
    cols = Array.from({ length: 9 }, () => new Set());
    rows = Array.from({ length: 9 }, () => new Set());
    squares = Array.from({ length: 9 }, () => new Set());
}

function isValid(board) {
    if (board.length !== 9 || board[0].length !== 9) {
        return false;
    }

    resetSets();

    for (let r = 0; r < 9; r++) {
        for (let c = 0; c < 9; c++) {
            if (board[r][c] === 0) continue;

            if (
                rows[r].has(board[r][c]) ||
                cols[c].has(board[r][c]) ||
                squares[3 * Math.floor(r / 3) + Math.floor(c / 3)].has(board[r][c])
            ) {
                return false;
            }

            rows[r].add(board[r][c]);
            cols[c].add(board[r][c]);
            squares[3 * Math.floor(r / 3) + Math.floor(c / 3)].add(board[r][c]);
        }
    }

    return true;
}

function isComplete(board) {
    return board.every(row => row.every(cell => cell !== 0));
}

function findNextEmptyCell(board) {
    for (let r = 0; r < 9; r++) {
        for (let c = 0; c < 9; c++) {
            if (board[r][c] === 0) {
                return [r, c];
            }
        }
    }

    return [-1, -1];
}

function isValidMove(board, row, col, num) {
    return !(
        rows[row].has(num) ||
        cols[col].has(num) ||
        squares[3 * Math.floor(row / 3) + Math.floor(col / 3)].has(num)
    )
}

function showPopup(message) {
    const popup = document.getElementById('popup');
    const popupMessage = document.getElementById('popup-message');
    popupMessage.textContent = message;
    popup.style.display = 'block';
}

document.getElementById('popup-close').addEventListener('click', () => {
    const popup = document.getElementById('popup');
    popup.style.display = 'none';
});

function solveSudoku() {
    const board = [];
    const cells = sudoku_board.querySelectorAll('input');

    for (let row = 0; row < 9; row++) {
        const currentRow = [];
        for (let col = 0; col < 9; col++) {
            const idx = row * 9 + col;
            const rawValue = cells[idx].value.trim();
            let value;

            if (rawValue === '') {
                value = 0;
            } else {
                value = parseInt(rawValue);
                if (isNaN(value) || value < 1 || value > 9) {
                    showPopup(
                        'Invalid value at row ${row + 1}, column ${col + 1}. Please enter a number from 1 to 9.'
                    );
                    return null;
                }
            }

            currentRow.push(value);
        }
        board.push(currentRow);
    }

    if (!isValid(board)) {
        showPopup('Board is not valid. Please enter a valid Sudoku board.');
        return null;
    }

    return backtrack(board) ? board : null;
}

function backtrack(board) {
    if (isComplete(board)) {
        return true;
    }

    const [row, col] = findNextEmptyCell(board);

    for (let num = 1; num <= 9; num++) {
        if (isValidMove(board, row, col, num)) {
            board[row][col] = num;
            rows[row].add(num);
            cols[col].add(num);
            squares[3 * Math.floor(row / 3) + Math.floor(col / 3)].add(num);

            if (backtrack(board)) {
                return board;
            }

            // Backtrack
            board[row][col] = 0;
            rows[row].delete(num);
            cols[col].delete(num);
            squares[3 * Math.floor(row / 3) + Math.floor(col / 3)].delete(num);
        }
    }

    return false;
}

function solve() {
    const solution = solveSudoku();
    
    const parent = document.getElementById('sudoku-container');
    
    // clean up any previous solutions
    const oldSolution = document.getElementById('solved-grid');
    if (oldSolution) {
        oldSolution.remove();
    }

    // create new solution grid
    const gridDiv = document.createElement('div');
    gridDiv.className = 'sudoku-board';
    gridDiv.id = 'solved-grid';

    if (solution) {
        for (let i = 0; i < 81; i++) {
            const input = document.createElement('input');
            input.type = 'text';
            input.maxLength = 1;
            input.value = solution[Math.floor(i / 9)][i % 9];
            input.disabled = true;
            gridDiv.appendChild(input);
        }
        parent.appendChild(gridDiv);
    }
}